File: frontend/src/api/company.ts
import type { CompanyType } from './profile'
import { del, get, patch, post, put } from './client'

export type CompanyStatus =
  | 'watching'
  | 'applied'
  | 'interviewing'
  | 'offered'
  | 'rejected'
  | 'archived'

export type CompanySource = 'ai_recommend' | 'manual'

export interface Company {
  id: number
  user_id: number
  name: string
  address: string
  province?: string
  city?: string
  district?: string
  longitude: number
  latitude: number
  category?: CompanyType
  industry?: string
  status: CompanyStatus
  source: CompanySource
  ai_reason?: string
  note?: string
  created_at: string
  updated_at: string
}

export interface CompanyCreateInput {
  name: string
  address: string
  province?: string
  city?: string
  district?: string
  longitude: number
  latitude: number
  category?: CompanyType
  industry?: string
  status?: CompanyStatus
  source?: CompanySource
  ai_reason?: string
  note?: string
}

export type CompanyUpdateInput = Partial<CompanyCreateInput>

export interface CompanyListQuery {
  status?: CompanyStatus
  category?: CompanyType
  keyword?: string
  page?: number
  page_size?: number
}

export interface CompanyListResult {
  list: Company[]
  pagination: {
    page: number
    page_size: number
    total: number
    total_pages: number
  }
}

export const listCompanies = (q: CompanyListQuery = {}) =>
  get<CompanyListResult>('/companies', q as Record<string, unknown>)

export const getCompany = (id: number) => get<Company>(`/companies/${id}`)

export const createCompany = (body: CompanyCreateInput) => post<Company>('/companies', body)

export const updateCompany = (id: number, body: CompanyUpdateInput) =>
  put<Company>(`/companies/${id}`, body)

export const updateCompanyStatus = (id: number, status: CompanyStatus) =>
  patch<Company>(`/companies/${id}/status`, { status })

export const deleteCompany = (id: number) => del<{ id: number }>(`/companies/${id}`)

export const batchCreateCompanies = (companies: CompanyCreateInput[]) =>
  post<{
    created: Company[]
    skipped: { name: string; reason: string }[]
    warning?: string
  }>('/companies/batch', { companies })

File: backend/internal/model/company.go
package model

import "time"

type Company struct {
	ID        int64     `json:"id"`
	UserID    int64     `json:"user_id"`
	Name      string    `json:"name"`
	Address   string    `json:"address"`
	Province  *string   `json:"province,omitempty"`
	City      *string   `json:"city,omitempty"`
	District  *string   `json:"district,omitempty"`
	Longitude float64   `json:"longitude"`
	Latitude  float64   `json:"latitude"`
	Category  *string   `json:"category,omitempty"`
	Industry  *string   `json:"industry,omitempty"`
	Status    string    `json:"status"`
	Source    string    `json:"source"`
	AIReason  *string   `json:"ai_reason,omitempty"`
	Note      *string   `json:"note,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type CompanyCreateInput struct {
	Name      string   `json:"name" binding:"required,max=128"`
	Address   string   `json:"address" binding:"required,max=512"`
	Province  *string  `json:"province" binding:"omitempty,max=32"`
	City      *string  `json:"city" binding:"omitempty,max=32"`
	District  *string  `json:"district" binding:"omitempty,max=32"`
	Longitude float64  `json:"longitude" binding:"required,gte=-180,lte=180"`
	Latitude  float64  `json:"latitude" binding:"required,gte=-90,lte=90"`
	Category  *string  `json:"category" binding:"omitempty,oneof=big_tech mid_tech startup foreign other"`
	Industry  *string  `json:"industry" binding:"omitempty,max=64"`
	Status    *string  `json:"status" binding:"omitempty,oneof=watching applied interviewing offered rejected archived"`
	Source    *string  `json:"source" binding:"omitempty,oneof=ai_recommend manual"`
	AIReason  *string  `json:"ai_reason"`
	Note      *string  `json:"note"`
}

type CompanyUpdateInput struct {
	Name      *string  `json:"name" binding:"omitempty,max=128"`
	Address   *string  `json:"address" binding:"omitempty,max=512"`
	Province  *string  `json:"province" binding:"omitempty,max=32"`
	City      *string  `json:"city" binding:"omitempty,max=32"`
	District  *string  `json:"district" binding:"omitempty,max=32"`
	Longitude *float64 `json:"longitude" binding:"omitempty,gte=-180,lte=180"`
	Latitude  *float64 `json:"latitude" binding:"omitempty,gte=-90,lte=90"`
	Category  *string  `json:"category" binding:"omitempty,oneof=big_tech mid_tech startup foreign other"`
	Industry  *string  `json:"industry" binding:"omitempty,max=64"`
	Status    *string  `json:"status" binding:"omitempty,oneof=watching applied interviewing offered rejected archived"`
	Note      *string  `json:"note"`
}

type CompanyStatusInput struct {
	Status string `json:"status" binding:"required,oneof=watching applied interviewing offered rejected archived"`
}

type CompanyListQuery struct {
	Status   *string `form:"status"`
	Category *string `form:"category"`
	Keyword  *string `form:"keyword"`
	Page     int     `form:"page,default=1" binding:"gte=1"`
	PageSize int     `form:"page_size,default=50" binding:"gte=1,lte=100"`
}

type CompanyListResult struct {
	List       []Company  `json:"list"`
	Pagination Pagination `json:"pagination"`
}

type Pagination struct {
	Page       int   `json:"page"`
	PageSize   int   `json:"page_size"`
	Total      int64 `json:"total"`
	TotalPages int   `json:"total_pages"`
}

type CompanyBatchInput struct {
	Companies []CompanyCreateInput `json:"companies" binding:"required,min=1,max=50,dive"`
}

type CompanyBatchResult struct {
	Created []Company        `json:"created"`
	Skipped []SkippedCompany `json:"skipped"`
	Warning *string          `json:"warning,omitempty"`
}

type SkippedCompany struct {
	Name   string `json:"name"`
	Reason string `json:"reason"`
}
